README.md
# telecom-universe-components

> OVH Manager Telecom Universe Components

[![Downloads](https://badgen.net/npm/dt/@ovh-ux/telecom-universe-components)](https://npmjs.com/package/@ovh-ux/telecom-universe-components) [![Dependencies](https://badgen.net/david/dep/ovh-ux/telecom-universe-components)](https://npmjs.com/package/@ovh-ux/telecom-universe-components?activeTab=dependencies) [![Dev Dependencies](https://badgen.net/david/dev/ovh-ux/telecom-universe-components)](https://npmjs.com/package/@ovh-ux/telecom-universe-components?activeTab=dependencies) [![Gitter](https://badgen.net/badge/gitter/ovh-ux/blue?icon=gitter)](https://gitter.im/ovh/ux)

We extracted all the components of the [telecom control panel](https://github.com/ovh-ux/ovh-manager-telecom) in order to interconnect them
both in our upcoming [monorepo](https://github.com/ovh-ux/manager) but also in the current stack.

At the end we are planning remove this repository and privilege the management of sources directly
inside the monorepo.

## Install

```sh
yarn add @ovh-ux/telecom-universe-components
```
## Usage

```js
import angular from 'angular';
import telecomUniverseComponents from '@ovh-ux/telecom-universe-components';

angular
  .module('myApp', [
    telecomUniverseComponents,
  ]);
```

## Test

```sh
yarn test
```

## Related

* [ovh-manager-telecom](https://github.com/ovh-ux/ovh-manager-telecom) - OVH Control Panel Telecom UI

## Contributing

Always feel free to help out! Whether it's [filing bugs and feature requests](https://github.com/ovh-ux/telecom-universe-components/issues/new) or working on some of the [open issues](https://github.com/ovh-ux/telecom-universe-components/issues), our [contributing guide](CONTRIBUTING.md) will help get you started.

## License

[BSD-3-Clause](LICENSE) © OVH SAS


CHANGELOG.md
## [3.0.8](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v3.0.7...v3.0.8) (2019-08-19)



## [3.0.7](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v3.0.6...v3.0.7) (2019-07-30)



## [3.0.6](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v3.0.5...v3.0.6) (2019-07-26)


### Bug Fixes

* **validator:** bump to version 11 ([#112](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/112)) ([04f5704](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/04f5704)), closes [validatorjs/validator.js#525](https://github.com/validatorjs/validator.js/issues/525)



## [3.0.5](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v3.0.4...v3.0.5) (2019-07-16)



## [3.0.4](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v3.0.3...v3.0.4) (2019-07-11)


### Bug Fixes

* **telephony.calls.filtering:** import missing module dependency ([23dbe3f](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/23dbe3f))



## [3.0.3](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v3.0.2...v3.0.3) (2019-05-10)


### Bug Fixes

* bump peer dep and replace deprecated method ([#93](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/93)) ([aca7af7](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/aca7af7))
* **i18n:** retrieve translations ([#95](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/95)) ([f82fa6e](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/f82fa6e))



## [3.0.2](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v3.0.1...v3.0.2) (2019-04-23)


### Bug Fixes

* **deps:** upgrade @ovh-ux/ng-ovh-mondial-relay to v6.0.1 ([#90](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/90)) ([50255ca](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/50255ca))



## [3.0.1](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v3.0.0...v3.0.1) (2019-04-23)


### Bug Fixes

* **shipping.mode.selection:** use component instead of directive ([#86](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/86)) ([8c90b5d](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/8c90b5d))



# [3.0.0](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v2.0.4...v3.0.0) (2019-03-12)


### Build System

* **peerdeps:** upgrade ng-ovh-swimming-poll to v4.0.0 ([#82](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/82)) ([c13aa3a](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/c13aa3a))


### BREAKING CHANGES

* **peerdeps:** replace `ovh-angular-swimming-poll` by `@ovh-ux/ng-ovh-swimming-poll`



## [2.0.4](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v2.0.3...v2.0.4) (2019-03-04)


### Bug Fixes

* **resiliation:** use the correct name for survey ([#74](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/74)) ([7d85373](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/7d85373))



## [2.0.3](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v2.0.2...v2.0.3) (2019-03-01)


### Bug Fixes

* **voip.alias:** add missing queue edit route parameter ([#75](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/75)) ([4e7c962](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/4e7c962))



## [2.0.2](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v2.0.1...v2.0.2) (2019-02-14)


### Bug Fixes

* **resiliation:** use the right binding for reason model ([#69](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/issues/69)) ([c54c55c](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/c54c55c))



## [2.0.1](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v2.0.0...v2.0.1) (2019-01-22)



# [2.0.0](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v1.5.1...v2.0.0) (2019-01-21)


### Code Refactoring

* add missing ovh prefix to the angular module name ([8539e22](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/8539e22))


### BREAKING CHANGES

* module is now named as `ngOvhTelecomUniverseComponents`



## [1.5.1](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v1.5.0...v1.5.1) (2019-01-09)


### Bug Fixes

* **i18n:** retrieve translations ([bd25c4c](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/bd25c4c))
* **toasterror:** add missing translation ([e9c5e5d](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/e9c5e5d))



# [1.5.0](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v1.4.0...v1.5.0) (2019-01-07)


### Features

* **voip.alias:** handle choice feature type parameter ([31a0a6a](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/31a0a6a))
* **voip.line:** add get phone method ([b2b0325](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/b2b0325))



# [1.4.0](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v1.3.0...v1.4.0) (2018-12-31)


### Features

* **bank-holidays:** add service to retrieve bank holidays ([b4b2b71](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/b4b2b71))



# [1.3.0](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v1.2.1...v1.3.0) (2018-12-07)


### Features

* **telecom.voip:** add missing voip alias feature ([0c05946](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/0c05946))



## [1.2.1](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/compare/v1.2.0...v1.2.1) (2018-12-07)


### Bug Fixes

* **pack.migration:** use offer description instead of description  ([c5d9432](https://github.com/ovh-ux/ng-ovh-telecom-universe-components/commit/c5d9432))

src/validator/index.js
import angular from 'angular';
import _ from 'lodash';
import punycode from 'punycode';
import validator from 'validator-js';

const moduleName = 'tucValidator';

angular
  .module(moduleName, [])
  .constant('tucValidator', validator)
  .run((tucValidator) => {
    const Rio = function Rio(rioStr) {
      if (!/[A-Z0-9+]{12}/i.test(rioStr)) {
        return;
      }
      const rioUp = rioStr.toUpperCase();
      this.isMobile = /^d{2}$/.test(rioUp.substr(0, 2)); // mobile RIO : only figures
      this.isFix = /^[F-Z][0-9A-Z]/.test(rioUp.substr(0, 2)); // fix RIO
      this.provider = rioUp.substr(0, 2);
      this.contractType = rioUp.substr(2, 1);
      this.customer = rioUp.substr(3, 6);
      this.control = rioUp.substr(9, 3);
      this.rio = rioUp;
    };
    const customValidator = tucValidator;

    function checkRioSum(str, control) {
      const order = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+';
      let a = 0;
      let b = 0;
      let c = 0;
      for (let i = 0; i < 19; i += 1) {
        const position = order.indexOf(str[i]);
        a = ((1 * a) + position) % 37;
        b = ((2 * b) + position) % 37;
        c = ((4 * c) + position) % 37;
      }
      return (order[a] + order[b] + order[c]) === control;
    }

    Rio.prototype.check = function check(phoneNumber) {
      const tel = (phoneNumber || '').replace(/[^0-9]/g, '');
      const rioValid = checkRioSum(
        this.provider + this.contractType + this.customer + tel,
        this.control,
      );
      const rioCompliant = (/^0[6-7]/.test(tel) && this.isMobile) || (/^0[1-5]/.test(tel) && this.isFix);
      return !!this.rio && rioValid && rioCompliant;
    };

    /**
     * Validate a RIO
     * @param {String} rio          RIO code
     * @param {String} phoneNumber  Corresponding phone number
     */
    customValidator.tucIsRio = (rio, phoneNumber) => (new Rio(rio)).check(phoneNumber);

    /**
     * Validate a Standard insee
     * @param {Object} obj          Object containing in
     */
    customValidator.tucHasInseeCode = obj => _.isObject(obj) && !!obj.inseeCode;

    /**
     * Validate a zipcode
     * @param {String} zipcode Zip code to validate
     * @param  {Array} filter  scope of the zip code
     */
    customValidator.tucIsZipcode = (zipcode, filter) => {
      const check = {
        frenchOverseas: /^9[78]\d{3}$/.test(zipcode),
        metropolitanFrance: /^\d{5}$/.test(zipcode) && (zipcode >= 1000) && (zipcode < 96000),
        fr: (/^\d{5}$/.test(zipcode) && (zipcode >= 1000) && (zipcode < 96000)) || /^9[78]\d{3}$/.test(zipcode),
        be: /^\d{4}$/.test(zipcode) && (zipcode >= 1000),
        de: /^\d{5}$/.test(zipcode) && (zipcode >= 10000),
        ch: /^\d{4}$/.test(zipcode) && (zipcode >= 1000) && (zipcode < 9499),
        es: /^\d{5}$/.test(zipcode) && (((zipcode >= 1000) && (zipcode < 53000)) || /^070/.test(zipcode) || /^071/.test(zipcode) || /^080/.test(zipcode)),

        // No real validation rules ! there are cases for London (new rule),
        // London (old rule), Eire, the rest of the country
        gb: true,
      };
      const theFilter = filter || Object.keys(check);
      return theFilter.reduce((all, thisTest) => all || check[thisTest], false);
    };

    /**
     *  Validate an IPv4Block or IPv6Block
     *  @param {String} str     IP representation string
     *  @param {Number} version IP version
     */
    customValidator.tucIsIPBlock = (str, version) => {
      if (version === 4 || version === 6) {
        const split = str.split('/');
        return split.length === 2 && customValidator.isIP(split[0], version)
          && parseInt(split[1], 10) > 0 && parseInt(split[1], 10) <= (version === 4 ? 32 : 128);
      }
      return customValidator.tucIsIPBlock(str, 4) || customValidator.tucIsIPBlock(str, 6);
    };

    /**
     *  Validate a private IP range (according to RFC 1918)
     *  @param {String} str     IP representation string
     *  @return {Boolean}
     */
    customValidator.tucIsPrivateIPv4 = (str) => {
      if (customValidator.isIP(str, 4)) {
        const ipBlock = str.split('.');
        const bitBlock = parseInt(ipBlock[0], 10);
        const secondBitBlock = parseInt(ipBlock[1], 10);
        switch (bitBlock) {
          case 192: // 16-bit block
            return secondBitBlock === 168;
          case 172: // 20-bit block
            return secondBitBlock >= 16 && secondBitBlock <= 31;
          default: // 24-bit block
            return parseInt(bitBlock, 10) === 10;
        }
      }

      return false;
    };

    /**
     *  Validate a siret code
     *  @param {String} siret SIRET code
     */
    customValidator.tucIsSiret = (siret) => {
      if (_.isEmpty(siret)) {
        return true;
      }
      const theSiret = `${/^\d{14}$/.test(siret) ? siret : '00000000000001'}`;
      const luhn = theSiret.split('').reverse().map((val, index) => (index % 2 ? val * 2 : parseInt(val, 10)));
      const luhnSum = luhn.reduce(
        (total, elt) => total + (elt < 10 ? elt : elt - 9),
        0,
      );
      return luhnSum % 10 === 0;
    };

    customValidator.tucIsFrenchLandLine = phone => /^0[1-5]([\\s\\-]?([0-9]){2}){4}$/.test(phone);

    customValidator.tucIsFrenchPhoneNumber = phone => /^(0033|\+33\s?(\(0\))?|0)[^08](\s*\d{2}){4}$/.test(phone);

    customValidator.tucIsMacAddress = (val) => {
      const values = val.split(/:/);
      return values.length === 6 && _.reduce(values, (result, elt) => /^[0-9a-f]{2}$/i.test(elt) && result, true);
    };

    /**
     * Check if a domain is valid
     * @param {String}  domain                          - The domain to check validity
     * @param {Object}  options                         - Options for validation check
     * @param {Boolean} options.canBeginWithUnderscore  - specifics NDD can be like:
                                                          _foo._bar.example.com
     * @param {Boolean} options.canBeginWithWildcard    - specifics NDD can be like:
                                                          *.foo.bar.example.com
     * @return {Boolean}
     */
    customValidator.tucIsValidDomain = (domain, options) => {
      const theOptions = options || {};
      let inError = false;

      if (domain) {
        const punycodeVersion = punycode.toASCII(domain.trim());
        const dotSplit = punycodeVersion.split('.');

        // Check lengths
        inError = punycodeVersion.length > 255 || dotSplit.length < 2;

        // Check wildcard
        if (!inError && punycodeVersion.indexOf('*') !== -1
          && (theOptions.canBeginWithWildcard ? !/^(?:\*\.)[^*]+$/.test(punycodeVersion) : true)) {
          inError = true;
        }

        // Check subdomain(s)
        if (!inError) {
          angular.forEach(dotSplit, (sub) => {
            if (sub.length > 63 || /(?:(?:^\s*$)|(?:^-)|(?:-$))/.test(sub)) {
              inError = true;
            }
            if (sub.indexOf('_') !== -1 && (theOptions.canBeginWithUnderscore ? !/^_[^_]+$/.test(sub) : true)) {
              inError = true;
            }
          });
        }

        // Check if it's not an IP
        if (!inError && customValidator.isIP(domain)) {
          inError = true;
        }

        // Check chars globally
        if (!inError) {
          inError = !/^[\w\-.*]+$/.test(punycodeVersion);
        }
      }
      return !inError;
    };
  });

export default moduleName;
